CSJ pruned_transducer_stateless7_streaming

teowenshen opened the pull request:
## CERs

The CERs are: 

#### Trained and evaluated on disfluent transcript
| decoding method	         | chunk size | eval1 | eval2 | eval3 | excluded | valid | average     | decoding mode       |
| -------------------------- | ---------- | ----- | ----- | ----- | -------- | ----- | ------------------- | ------------------- |
| fast beam search           | 320ms	  | 6.27  | 5.13  | 5.05  | 6.30     | 5.42  | --epoch 30 --avg 14 | simulated streaming |
| fast beam search           | 320ms	  | 5.91  | 4.30  | 4.53  | 6.13     | 5.13  | --epoch 30 --avg 14 | chunk-wise          |
| greedy search              | 320ms      | 5.81  | 4.29  | 4.63  | 6.07     | 5.13  | --epoch 30 --avg 14 | simulated streaming |
| greedy search              | 320ms      | 5.91  | 4.50  | 4.65  | 6.36     | 5.34  | --epoch 30 --avg 14 | chunk-wise          |
| modified beam search       | 320ms      | 5.59  | 4.20  | 4.39  | 5.54     | 4.90  | --epoch 30 --avg 14 | simulated streaming |
| modified beam search       | 320ms      | 5.79  | 4.48  | 4.41  | 5.98     | 5.19  | --epoch 30 --avg 14 | chunk-wise          |
| fast beam search           | 640ms      | 5.76  | 4.35  | 4.39  | 5.40     | 4.92  | --epoch 30 --avg 14 | simulated streaming |
| fast beam search           | 640ms      | 5.45  | 4.31  | 4.29  | 5.61     | 4.97  | --epoch 30 --avg 14 | chunk-wise          |
| greedy search              | 640ms      | 5.37  | 3.94  | 4.03  | 5.22     | 4.77  | --epoch 30 --avg 14 | simulated streaming |
| greedy search              | 640ms      | 5.77  | 4.44  | 4.49  | 5.70     | 5.29  | --epoch 30 --avg 14 | chunk-wise          |
| modified beam search       | 640ms      | 5.19  | 3.81  | 3.93  | 4.83     | 4.59  | --epoch 30 --avg 14 | simulated streaming |
| modified beam search       | 640ms      | 6.71  | 5.35  | 4.95  | 6.06     | 5.94  | --epoch 30 --avg 14 | chunk-wise  |

#### Trained and evaluated on fluent transcript
| decoding method  |  chunk size  |  eval1  |  eval2  |  eval3  |  excluded  |  valid  |  average  |  decoding mode |
| ---------------  | ------------ | ------- | ------ | ------- | ----------- | ------- | --------- | -------------- |
| fast beam search pad30  |  320ms  |  4.72  |  3.74  |  4.21  |  5.21  |  4.39  |  --epoch 30 --avg 19  |  simulated streaming |
| fast beam search  |  320ms  |  4.63  |  3.63  |  4.18  |  5.3  |  4.31  |  --epoch 30 --avg 19  |  chunk-wise |
| greedy search  |  320ms  |  4.83  |  3.71  |  4.27  |  4.89  |  4.38  |  --epoch 30 --avg 19  |  simulated streaming |
| greedy search  |  320ms  |  4.7  |  3.87  |  4.24  |  5.39  |  4.39  |  --epoch 30 --avg 19  |  chunk-wise |
| modified beam search  |  320ms  |  4.61  |  3.55  |  4.07  |  4.89  |  4.18  |  --epoch 30 --avg 19  |  simulated streaming |
| modified beam search  |  320ms  |  4.53  |  3.73  |  3.98  |  5.9  |  4.25  |  --epoch 30 --avg 19  |  chunk-wise |
| fast beam search pad30  |  640ms  |  4.33  |  3.55  |  4.03  |  4.97  |  4.33  |  --epoch 30 --avg 19  |  simulated streaming |
| fast beam search  |  640ms  |  4.21  |  3.64  |  3.93  |  5.04  |  4.18  |  --epoch 30 --avg 19  |  chunk-wise |
| greedy search  |  640ms  |  4.3  |  3.51  |  3.91  |  4.45  |  4.04  |  --epoch 30 --avg 19  |  simulated streaming |
| greedy search  |  640ms  |  4.4  |  3.83  |  4.03  |  5.14  |  4.31  |  --epoch 30 --avg 19  |  chunk-wise |
| modified beam search  |  640ms  |  4.11  |  3.29  |  3.66  |  4.33  |  3.88  |  --epoch 30 --avg 19  |  simulated streaming |
| modified beam search  |  640ms  |  4.42  |  3.91  |  3.93  |  5.62  |  4.33  |  --epoch 30 --avg 19  |  chunk-wise |

#### Comparing disfluent and fluent models 
$$\texttt{CER}^{f}_d = \frac{\texttt{sub}_f + \texttt{ins} + \texttt{del}_f}{N_f}$$

This comparison evaluates the disfluent model on the fluent transcript (calculated by `disfluent_recogs_to_fluent.py`), forgiving the disfluent model's mistakes on fillers and partial words. It is meant as an illustrative metric only, so that the disfluent and fluent models can be compared. 

| decoding method | chunk size | eval1 (d vs f) | eval2 (d vs f) | eval3  (d vs f) | excluded  (d vs f) | valid  (d vs f) | decoding mode | 
| --------------- | ---------- | -------------- | -------------- | --------------- | ------------------ | --------------- | ------------- |
| fast beam search | 320ms | 5.44 vs 4.72 | 4.49 vs 3.74 | 4.44 vs 4.21 | 5.14 vs 5.21 | 4.64 vs 4.39 | simulated streaming | 
| fast beam search | 320ms | 5.05 vs 4.63 | 3.63 vs 3.63 | 3.91 vs 4.18 | 4.75 vs 5.30 | 4.29 vs 4.31 | chunk-wise | 
| greedy search | 320ms | 4.97 vs 4.83 | 3.63 vs 3.71 | 4.02 vs 4.27 | 4.93 vs 4.89 | 4.32 vs 4.38 | simulated streaming | 
| greedy search | 320ms | 5.02 vs 4.70 | 3.78 vs 3.87 | 4.02 vs 4.24 | 5.11 vs 5.39 | 4.47 vs 4.39 | chunk-wise  | 
| modified beam search | 320ms | 4.86 vs 4.61 | 3.62 vs 3.55 | 3.85 vs 4.07 | 4.66 vs 4.89 | 4.21 vs 4.18 | simulated streaming | 
| modified beam search | 320ms | 5.05 vs 4.53 | 3.89 vs 3.73 | 3.88 vs 3.98 | 4.88 vs 5.90 | 4.48 vs 4.25 | chunk-wise | 
| fast beam search | 640ms | 4.93 vs 4.33 | 3.74 vs 3.55 | 3.78 vs 4.03 | 4.31 vs 4.97 | 4.15 vs 4.33 | simulated streaming | 
| fast beam search | 640ms | 4.61 vs 4.21 | 3.67 vs 3.64 | 3.66 vs 3.93 | 4.34 vs 5.04 | 4.15 vs 4.18 | chunk-wise | 
| greedy search | 640ms | 4.48 vs 4.30 | 3.29 vs 3.51 | 3.43 vs 3.91 | 4.11 vs 4.45 | 3.96 vs 4.04 | simulated streaming | 
| greedy search | 640ms | 4.89 vs 4.40 | 3.77 vs 3.83 | 3.87 vs 4.03 | 4.41 vs 5.14 | 4.47 vs 4.31 | chunk-wise | 
| modified beam search | 640ms | 4.45 vs 4.11 | 3.28 vs 3.29 | 3.41 vs 3.66 | 3.97 vs 4.33 | 3.90 vs 3.88 | simulated streaming | 
| modified beam search | 640ms | 6.10 vs 4.42 | 4.86 vs 3.91 | 4.51 vs 3.93 | 5.16 vs 5.62 | 5.34 vs 4.33 | chunk-wise | 
| __average of (d - f)__ |  | 0.50 | 0.14 | -0.13 | -0.45 | 0.11 | |

## Commands for disfluent training and decoding

The training command was: 
```bash
./pruned_transducer_stateless7_streaming/train.py \
  --feedforward-dims  "1024,1024,2048,2048,1024" \
  --context-size 2 \
  --world-size 8 \
  --num-epochs 30 \
  --start-epoch 1 \
  --use-fp16 1 \
  --exp-dir pruned_transducer_stateless7_streaming/exp_disfluent \
  --max-duration 375 \
  --transcript-mode disfluent \
  --lang data/lang_char \
  --musan-dir /mnt/host/corpus/musan/musan/fbank
```

Padding with 30 caused many insertions at the end of utterances. The simulated streaming decoding command was:
```bash
for chunk in 64 32; do
    for m in greedy_search fast_beam_search modified_beam_search; do
        python pruned_transducer_stateless7_streaming/decode.py \
            --feedforward-dims  "1024,1024,2048,2048,1024" \
            --exp-dir pruned_transducer_stateless7_streaming/exp_disfluent_2 \
            --epoch 30 \
            --avg 14 \
            --max-duration 250 \
            --decoding-method $m \
            --manifest-dir /mnt/host/corpus/csj/fbank \
            --lang data/lang_char \
            --transcript-mode disfluent \
            --res-dir pruned_transducer_stateless7_streaming/exp_disfluent_2/github/sim_"$chunk"_"$m" \
            --decode-chunk-len $chunk \
            --pad 4
        done
    done
done
```

The streaming chunk-wise decoding command was:
```bash
for chunk in 64 32; do
    for m in greedy_search fast_beam_search modified_beam_search; do
        python pruned_transducer_stateless7_streaming/streaming_decode.py \
            --feedforward-dims  "1024,1024,2048,2048,1024" \
            --exp-dir pruned_transducer_stateless7_streaming/exp_disfluent_2 \
            --epoch 30 \
            --avg 14 \
            --max-duration 250 \
            --decoding-method $m \
            --manifest-dir /mnt/host/corpus/csj/fbank \
            --lang data/lang_char \
            --transcript-mode disfluent \
            --res-dir pruned_transducer_stateless7_streaming/exp_disfluent_2/github/stream_"$chunk"_"$m" \
            --decode-chunk-len $chunk \
            --num-decode-streams 40
    done
done
```

## Commands for fluent training and decoding
The training command was: 
```bash
./pruned_transducer_stateless7_streaming/train.py \
  --feedforward-dims  "1024,1024,2048,2048,1024" \
  --context-size 2 \
  --world-size 8 \
  --num-epochs 30 \
  --start-epoch 1 \
  --use-fp16 1 \
  --exp-dir pruned_transducer_stateless7_streaming/exp_fluent_2 \
  --max-duration 375 \
  --transcript-mode fluent \
  --telegram-cred misc.ini \
  --lang data/lang_char \
  --manifest-dir $csj_fbank_dir \
  --musan-dir /mnt/host/corpus/musan/musan/fbank
```

The simulated streaming decoding command was:
```bash
for chunk in 64 32; do
    for m in greedy_search modified_beam_search; do
        python pruned_transducer_stateless7_streaming/decode.py \
            --feedforward-dims  "1024,1024,2048,2048,1024" \
            --exp-dir pruned_transducer_stateless7_streaming/exp_fluent_2 \
            --epoch 30 \
            --avg 19 \
            --max-duration 350 \
            --decoding-method $m \
            --manifest-dir /mnt/host/corpus/csj/fbank \
            --lang data/lang_char \
            --transcript-mode fluent \
            --res-dir pruned_transducer_stateless7_streaming/exp_fluent_2/github/sim_"$chunk"_"$m" \
            --decode-chunk-len $chunk \
            --pad 4
    done
    # Padding of 4 caused many deletions only in the fast_beam_search case. 
    python pruned_transducer_stateless7_streaming/decode.py \
        --feedforward-dims  "1024,1024,2048,2048,1024" \
        --exp-dir pruned_transducer_stateless7_streaming/exp_fluent_2 \
        --epoch 30 \
        --avg 19 \
        --max-duration 350 \
        --decoding-method fast_beam_search \
        --manifest-dir /mnt/host/corpus/csj/fbank \
        --lang data/lang_char \
        --transcript-mode fluent \
        --res-dir pruned_transducer_stateless7_streaming/exp_fluent_2/github/sim_"$chunk"_fast_beam_search \
        --decode-chunk-len $chunk \
        --pad 30
done
```

The streaming chunk-wise decoding command was:
```bash
for chunk in 64 32; do
    for m in greedy_search fast_beam_search modified_beam_search; do
        python pruned_transducer_stateless7_streaming/streaming_decode.py \
            --feedforward-dims  "1024,1024,2048,2048,1024" \
            --exp-dir pruned_transducer_stateless7_streaming/exp_fluent_2 \
            --epoch 30 \
            --avg 19 \
            --max-duration 250 \
            --decoding-method $m \
            --manifest-dir /mnt/host/corpus/csj/fbank \
            --lang data/lang_char \
            --transcript-mode fluent \
            --res-dir pruned_transducer_stateless7_streaming/exp_fluent_2/github/stream_"$chunk"_"$m" \
            --decode-chunk-len $chunk \
            --gpu 4 \
            --num-decode-streams 40
    done
done
```

csukuangfj commented:
@teowenshen 

Thanks! Is it ready for review?

teowenshen commented:
I am writing up RESULTS.md and README.md now, but will publish them once I manage to upload my results and model to HuggingFace. Also, I still couldn't wrap my head around the difference between streaming_decode.py and decode.py. 

Other than that, this PR is ready.

pkufool commented:
> Also, I still couldn't wrap my head around the difference between streaming_decode.py and decode.py.

About the difference between streaming_decode.py and decode.py please see discussion [here](https://github.com/k2-fsa/icefall/discussions/744#discussioncomment-4338452)  and the documents [here](https://k2-fsa.github.io/icefall/recipes/Streaming-ASR/librispeech/pruned_transducer_stateless.html#decoding)

teowenshen commented:
Apologies! I meant, I couldn't find out why the results of `streaming_decode.py` and `decode.py` have such a CER gap, since given the same chunk length and left context they should be almost the same. I have a more detailed finding here: https://github.com/k2-fsa/icefall/issues/807, and have since then retrained another model to compare.

teowenshen commented:
I have retrained an early model with padding (30) so that when decoding with padding 30 the insertions at end of utterances are less. This improved the simulated streaming `decode.py` results, but still the chunk-wise `streaming_decode.py` results are bad. 

How can I pad the `streaming_decode.py` with 30 frames only, like `decode.py`?

csukuangfj commented:
I just updated
https://huggingface.co/spaces/k2-fsa/automatic-speech-recognition
to include pre-trained models from this pull-request.

![Screen Shot 2023-02-12 at 10 41 10](https://user-images.githubusercontent.com/5284924/218289921-fe8f75c5-eb05-49e6-8614-3e43e6fb5cc7.png)


---

The test wave is downloaded from
https://clrd.ninjal.ac.jp/csj/sound-f/aps-smp.mp3

The ground truth is
https://clrd.ninjal.ac.jp/csj/trans-f/aps-smp.txt

csukuangfj commented:
@teowenshen 

Could you please add a `test_wavs` folder to
https://huggingface.co/TeoWenShen/icefall-asr-csj-pruned-transducer-stateless7-streaming-230208/tree/main
?

You can use
https://huggingface.co/csukuangfj/icefall-asr-librispeech-pruned-transducer-stateless8-2022-11-14/tree/main/test_wavs
as a reference.

teowenshen commented:
@csukuangfj Thank you for your comments. I have added `test_wavs` to https://huggingface.co/TeoWenShen/icefall-asr-csj-pruned-transducer-stateless7-streaming-230208/tree/main .

csukuangfj commented on egs/csj/ASR/RESULTS.md:
```diff
@@ -0,0 +1,212 @@
+  --feedforward-dims  "1024,1024,2048,2048,1024" \
+  --context-size 2 \
+  --world-size 8 \
+  --num-epochs 30 \
```
Are the results listed in the table from the best combination `--epoch 30 --avg 14`? That is, have you tried other combinations of `--epoch --avg`?

teowenshen replied:
Yes, the results listed in the table are from an exhaustive search for the best combination, for both disfluent and fluent.

csukuangfj commented on egs/csj/ASR/local/disfluent_recogs_to_fluent.py:
```diff
@@ -0,0 +1,202 @@
+
+ARGPARSE_DESCRIPTION = """
+This helper code takes in a disfluent recogs file generated from icefall.utils.store_transcript,
+compares it against a fluent transcript, and save the results in a separate directory.
```
```suggestion
compares it against a fluent transcript, and saves the results in a separate directory.
```

csukuangfj commented on egs/csj/ASR/local/prepare_lang_char.py:
```diff
@@ -50,98 +40,58 @@ def get_args():
-        if args.trans_mode:
-            p += f"_{args.trans_mode}"
-        args.lang_dir = Path(p)
+    # if not args.lang_dir:
```
Could you remove it?

csukuangfj commented on egs/csj/ASR/local/utils/tokenizer.py:
```diff
@@ -0,0 +1,247 @@
+            "--lang-type",
+            type=str,
+            default=None,
+            help="Either 'bpe', 'char', or 'phone'. If not provided, it expects lang_dir/lang_type to exists.",
```
```suggestion
            help="Either 'bpe' or 'char'. If not provided, it expects lang_dir/lang_type to exists.",
```

Looks like `phone` is not supported at present.

teowenshen replied:
Thanks for the reminder. Currently this recipe does not support BPE too. I will make those changes.

csukuangfj commented on egs/csj/ASR/local/utils/tokenizer.py:
```diff
@@ -0,0 +1,247 @@
+        return self._id2sym[id]
+
+    def get_piece_size(self) -> int:
+        return len(self._sym2id)
```
I think `tokens.txt` contains disambig symbols, like `#0`, `#1`, etc.
Does this function also count disambig symbols?

teowenshen replied:
In this recipe, I simply loop through all transcript texts to get a unique set of output tokens, so there are no disambig symbols for homonyms nor backoff symbol in the token set.

teowenshen replied:
I just realised that all the Mandarin recipes generate disambig symbols in their `tokens.txt` too in `prepare_char.py`. 

Can I make these changes in a separate PR? I would need some time to implement these changes in this recipe.

csukuangfj replied:
> Can I make these changes in a separate PR?

Yes, a separate PR is fine.

Could you change
https://github.com/k2-fsa/icefall/blob/ba1505391e59a637d43d17018b69dba46643aa93/egs/csj/ASR/local/utils/tokenizer.py#L195-L196
to
```python3
 def get_piece_size(self) -> int: 
     assert '#0' not in self._sym2id
     return len(self._sym2id) 
```
so that we would not accidentally count disambig symbols.

teowenshen replied:
I added this check in the constructor of CharTokenizer.

csukuangfj reviewed (commented):
Thanks!

Left some minor comments.

teowenshen commented:
@csukuangfj I hope it's not too late, but actually I have much better results after padding both in training and decoding. I am updating the CER tables, the Huggingface repo, and the script commands. 

Code-wise, the only addition is a "--pad-feature" argument in `train.py`, so the existing bulk of codes that you've reviewed are mostly the same. So sorry for the hassle! 🙏

teowenshen commented:
@csukuangfj  Just to update that I have addressed your comments, and that this branch and the HuggingFace repo have been updated with the new results.

teowenshen commented:
@csukuangfj Yes. It is ready for merge from my end. Thanks!

csukuangfj commented:
@teowenshen 

Here is a demo on iPhone using the model trained from this pull-request.

It uses https://github.com/k2-fsa/sherpa-ncnn for deployment.

https://user-images.githubusercontent.com/5284924/219043902-46cf94f8-8699-41a0-bf44-1b8d6beee130.mp4

csukuangfj commented:
@teowenshen

teowenshen commented:
Yes, sure. `disfluent` is the transcript mode that most other CSJ ASR recipes are based on. `fluent` is where disfluency elements, like fillers and partial words, are removed. 

In the CSJ transcript, words are explicitly tagged to express a variety of information. You can refer table 3 of [this paper](http://www.lrec-conf.org/proceedings/lrec2000/pdf/262.pdf) to get an idea. In this recipe, the `fluent` model is trained on transcripts where all F-, D-, and D2-tags are removed, and all other tags are dealt with the same way as `disfluent`. 

(Kaldi is trained on the `disfluent` transcript. ESPnet uses the `disfluent` transcript too, if I'm not mistaken. Basically, for academic comparison purposes, we can use the `disfluent` transcript mode. However, because CSJ's original utterances are much shorter than 10s, this recipe's transcript is not exactly the same as Kaldi due to differences in the utterance concatenation algorithm.)

However, a problem I noticed early on while training the `disfluent` model is that many errors happen around fillers and partial words, and in fact those errors aren't quite important in the actual text. 

For example, 

| Type | Utt ID | Disfluent | Fluent | Remarks | 
| ---- | ------- | -------- | ------- | --------- | 
| Filler | A01M0097_0010 | [え (*->ー) こ の 図 は ...]( https://huggingface.co/TeoWenShen/icefall-asr-csj-pruned-transducer-stateless7-streaming-230208/blob/556e388c82be6dee687dae3e1385afff7784e1bc/decoding_results/disfluent/sim_64_modified_beam_search/errs-eval1-beam_size_4-epoch-30-avg-17-streaming-chunk-size-64-modified_beam_search-beam-size-4-use-averaged-model.txt#L10) | [こ の 図 は ...](https://huggingface.co/TeoWenShen/icefall-asr-csj-pruned-transducer-stateless7-streaming-230208/blob/556e388c82be6dee687dae3e1385afff7784e1bc/decoding_results/fluent/sim_64_modified_beam_search/errs-eval1-beam_size_4-epoch-30-avg-12-streaming-chunk-size-64-modified_beam_search-beam-size-4-use-averaged-model.txt#L10) | Error in the spelling of the filler by the `disfluent` model. | 
| Filler | A01M0097_0168 | [え ー (*->っ) と ー](https://huggingface.co/TeoWenShen/icefall-asr-csj-pruned-transducer-stateless7-streaming-230208/blob/556e388c82be6dee687dae3e1385afff7784e1bc/decoding_results/disfluent/sim_64_modified_beam_search/errs-eval1-beam_size_4-epoch-30-avg-17-streaming-chunk-size-64-modified_beam_search-beam-size-4-use-averaged-model.txt#L57) | [<blk>](https://huggingface.co/TeoWenShen/icefall-asr-csj-pruned-transducer-stateless7-streaming-230208/blob/556e388c82be6dee687dae3e1385afff7784e1bc/decoding_results/fluent/sim_64_modified_beam_search/errs-eval1-beam_size_4-epoch-30-avg-12-streaming-chunk-size-64-modified_beam_search-beam-size-4-use-averaged-model.txt#L57) | Whole utterance consists of a filler. The `disfluent` model wrongly spelled the filler, while the `fluent` model correctly ignored the filler. | 
| Partial word: Stuttering | A01M0097_0026 | [(ね こ し ゅ ぼ ん->と な っ て し ま う) え ー (*->っ) と こ こ で ...](https://huggingface.co/TeoWenShen/icefall-asr-csj-pruned-transducer-stateless7-streaming-230208/blob/556e388c82be6dee687dae3e1385afff7784e1bc/decoding_results/disfluent/sim_64_modified_beam_search/errs-eval1-beam_size_4-epoch-30-avg-17-streaming-chunk-size-64-modified_beam_search-beam-size-4-use-averaged-model.txt#L17) | [こ こ で ...](https://huggingface.co/TeoWenShen/icefall-asr-csj-pruned-transducer-stateless7-streaming-230208/blob/556e388c82be6dee687dae3e1385afff7784e1bc/decoding_results/fluent/sim_64_modified_beam_search/errs-eval1-beam_size_4-epoch-30-avg-12-streaming-chunk-size-64-modified_beam_search-beam-size-4-use-averaged-model.txt#L17 ) | The `disfluent` model wrongly spelled the gibberish partial word and the filler, while the `fluent` model correctly ignored both the partial word and the filler. |
| Partial word: Speech correction | A01M0110_0030 | [(に ほ ん->日 本) 難 易 度 ...](https://huggingface.co/TeoWenShen/icefall-asr-csj-pruned-transducer-stateless7-streaming-230208/blob/556e388c82be6dee687dae3e1385afff7784e1bc/decoding_results/disfluent/sim_64_modified_beam_search/errs-eval1-beam_size_4-epoch-30-avg-17-streaming-chunk-size-64-modified_beam_search-beam-size-4-use-averaged-model.txt#L124) | [(*->日 本) 難 易 度](https://huggingface.co/TeoWenShen/icefall-asr-csj-pruned-transducer-stateless7-streaming-230208/blob/556e388c82be6dee687dae3e1385afff7784e1bc/decoding_results/fluent/sim_64_modified_beam_search/errs-eval1-beam_size_4-epoch-30-avg-12-streaming-chunk-size-64-modified_beam_search-beam-size-4-use-averaged-model.txt#L124) | Because the partial word itself is a full word, the `disfluent` model spelled it with Chinese characters although the transcript used phonetic hiragana. The `fluent` model unfortunately picked up on the partial word too. |

So, I uploaded the `fluent` model too because, in my opinion, the `fluent` model is more immediately useful. Plus, the training codes are the same anyway.

csukuangfj commented:
@teowenshen 

Thanks for your detailed explanation!

csukuangfj commented:
@teowenshen 

Is the model at https://huggingface.co/TeoWenShen/icefall-asr-csj-pruned-transducer-stateless7-streaming-230208
suitable for commercial use?

If yes, could you add a README.md to it containing something like below:
```
---
license: apache-2.0
---
```

You can find an example at
https://huggingface.co/csukuangfj/sherpa-ncnn-conv-emformer-transducer-2022-12-06/raw/main/README.md

teowenshen commented:
I'm sorry, I've just clarified - The models from this pull request are not available for commercial use. 

I've taken down the models online. Can you help to delete the models from the demo page at https://huggingface.co/spaces/k2-fsa/automatic-speech-recognition too, since it is no longer working?

Since the original Huggingface link to the model has been deleted, I will send in a pull request to edit the README file. 

Deepest apologies for the inconvenience! I will try to attempt a recipe for another openly available Japanese corpus if time permits.

csukuangfj commented:
@teowenshen 
Sorry to hear that. I just removed it from https://huggingface.co/spaces/k2-fsa/automatic-speech-recognition

Hope that  a new Japanese model will be available soon.

---

By the way, I have converted this model to sherpa-ncnn and sherpa-onnx, shall I also take them down?

- https://huggingface.co/csukuangfj/sherpa-ncnn-streaming-zipformer-ja-fluent-2023-02-14
- https://huggingface.co/csukuangfj/sherpa-ncnn-streaming-zipformer-ja-disfluent-2023-02-14
- https://huggingface.co/csukuangfj/sherpa-ncnn-streaming-zipformer-ja-fluent-2023-02-14

csukuangfj commented:
@teowenshen 
Also, can I make the exported models private and use them only to demonstrate `k2-fsa`?

teowenshen commented:
> By the way, I have converted this model to sherpa-ncnn and sherpa-onnx, shall I also take them down?
> 
>     https://huggingface.co/csukuangfj/sherpa-ncnn-streaming-zipformer-ja-fluent-2023-02-14
>     https://huggingface.co/csukuangfj/sherpa-ncnn-streaming-zipformer-ja-disfluent-2023-02-14
>     https://huggingface.co/csukuangfj/sherpa-ncnn-streaming-zipformer-ja-fluent-2023-02-14

Yes, please help to take them down as well. Thanks!

> Also, can I make the exported models private and use them only to demonstrate `k2-fsa`?

I am really sorry as it is not my place to permit any use of derivatives from CSJ.  

I just want to say that I am truly on board with the open source spirit of k2-fsa / NGK, and personally have nothing against the commercial use of models. Once again, really sorry for the inconvenience!

csukuangfj commented:
@teowenshen 
Thanks, I see. I am taking them down.
